### ISSUE TITLE
NLog Exceptions don't get logged in NLog diagnostics

### ISSUE THREAD

**Issue by jamescrowley** [NLog/NLog#261]
When throwExceptions="true" is set on the NLog node, we could see the following:

    System.AggregateException: One or more errors occurred. ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> NLog.NLogRuntimeException: Exception occurred in NLog ---> System.ArgumentException: The specified string is not in the form required for a subject.
       at System.Net.Mail.Message.set_Subject(String value)
       at NLog.Targets.MailTarget.SetupMailMessage(MailMessage msg, LogEventInfo logEvent)
       at NLog.Targets.MailTarget.ProcessSingleMailMessage(List`1 events)
       --- End of inner exception stack trace ---
       at NLog.LoggerImpl.<>c__DisplayClass1.<Write>b__0(Exception ex)
       at NLog.Internal.SingleCallContinuation.Function(Exception exception)
       at NLog.Targets.MailTarget.ProcessSingleMailMessage(List`1 events)
       at NLog.Targets.MailTarget.Write(AsyncLogEventInfo[] logEvents)
       at NLog.Targets.Target.WriteAsyncLogEvent(AsyncLogEventInfo logEvent)
       at NLog.LoggerImpl.WriteToTargetWithFilterChain(TargetWithFilterChain targetListHead, LogEventInfo logEvent, AsyncContinuation onException)
       at NLog.LoggerImpl.Write(Type loggerType, TargetWithFilterChain targets, LogEventInfo logEvent, LogFactory factory)

However, with nLog set to the following:

     <nlog autoReload="true" internalLogFile="nlog.log" internalLogLevel="Trace">

You don't see any error occurring at all in the log file. I'd have expected an Error to be logged here?

**Comment by 304NotModified** [comment]
@Xharze your advise please.
I See in the code:

```
      catch (Exception exception)
            {
                if (exception.MustBeRethrown())
                {
                    throw;
                }

                InternalLogger.Error(exception.ToString());
```

So if `exception.MustBeRethrown == true`, it's never logged. Is this intended behavior?

**Comment by Xharze** [comment]
I can understand the reasoning for not logging the exception when it's
thrown, but I also think it would be nice to log the error in the internal
log. It will make it easier to locate issues when just looking at the log
file.

**Comment by ilya-g** [comment]
In case of exception is a `StackOverflowException` or `ThreadAbortException` or `OutOfMemoryException`, it should be rethrown as soon as possible, because doing some other calls in such extreme conditions can cause other exceptions to be thrown, which in the `catch` block will effectively mask the original exception and make it hard to diagnose the cause of it.

**Comment by 304NotModified** [comment]
This has been fixed in `MailTarget` with commit https://github.com/NLog/NLog/commit/0601c07ac298dc82408597af8c4ffb928a3eaa15

**Comment by 304NotModified** [comment]
this is structural fixed in #1143
